3.2. Линейная регрессия

In [227]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [228]:
myData = pd.read_csv('data/mycar.csv')

In [229]:
X = myData.iloc[:,:-1].values
Y = myData.iloc[:,1].values

In [230]:
from sklearn.model_selection import train_test_split

In [231]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size= 0.3)

In [232]:
from sklearn.linear_model import LinearRegression
myModel = LinearRegression() #Обозначаем, что наша модель - линейная регрессия
myModel.fit(X_train,Y_train) #обучаем модель на обучающих данных

LinearRegression()

In [233]:
y_pred = myModel.predict(X_test)
y_pred

array([34.31953578, 47.91644101, 27.52108317, 58.11411992, 41.11798839,
       13.92417795, 51.31566731, 34.31953578, 58.11411992, 27.52108317,
       37.71876209, 30.92030948, 41.11798839, 75.11025145, 24.12185687])

3.5. Линейная регрессия. Практика №1

In [234]:
#
import numpy as np
import pandas as pd
from sklearn.datasets import load_boston
from sklearn.datasets import fetch_california_housing
from sklearn.metrics import mean_squared_error, f1_score, accuracy_score, roc_curve, roc_auc_score
from sklearn.model_selection import train_test_split

from matplotlib import pyplot as plt

In [235]:
#certificates
import ssl
ssl._create_default_https_context = ssl._create_unverified_context

In [236]:
#Theta function
def linreg_linear(X, y):
    xTrans = X.T
    matrix = xTrans.dot(X)
    matrix = np.linalg.inv(matrix)
    matrix = matrix.dot(xTrans)
    matrix = matrix.dot(y)
    return matrix

In [265]:
#Подготавливаем данные и вычисляем параметр :
data = load_boston()
dataCalifornia = fetch_california_housing()
X, y = data['data'], data['target']
# create ones array  [1,1,1,1 ...] from X rows and create from it new column 
e = np.ones(X.shape[0])[:, np.newaxis]
# Join a sequence of arrays along a new axis.
X = np.hstack([e, X])

theta = linreg_linear(X, y)

In [238]:
theta

array([ 3.64594884e+01, -1.08011358e-01,  4.64204584e-02,  2.05586264e-02,
        2.68673382e+00, -1.77666112e+01,  3.80986521e+00,  6.92224640e-04,
       -1.47556685e+00,  3.06049479e-01, -1.23345939e-02, -9.52747232e-01,
        9.31168327e-03, -5.24758378e-01])

In [239]:
lenth_vect = 0
for x in theta:
    lenth_vect += x*x
print(lenth_vect**0.5)

40.867455573954125


In [202]:
theta.shape

(14,)

In [203]:
np.linalg.norm(theta) 

40.867455573954125

In [204]:
print(f'Ymax {y.max()}, Xmax {X.max()}')

Ymax 50.0, Xmax 711.0


In [205]:
X.shape

(506, 14)

In [31]:
np.ones(X.shape[0])[:, np.newaxis]

array([[1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],
       [1.],

In [35]:
np.ones(X.shape[0])[np.newaxis, :]

array([[1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.,
        1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 

In [206]:
# Сделаем предсказание и посчитаем метрики:
y_pred = X.dot(theta)

def print_regression_metrics(y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    print(f'MSE = {mse:.2f}, RMSE = {rmse:.2f}')


In [207]:
#для полноразмерного обучения -полной выборки
print_regression_metrics(y, y_pred)

MSE = 21.89, RMSE = 4.68


In [208]:
#Давайте разобьём выборку на train/valid, вычислим , сделаем предсказания и посчитаем ошибки MSE и RMSE:
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2)
theta = linreg_linear(X_train, y_train)
y_valid_pred = X_valid.dot(theta)
y_train_pred = X_train.dot(theta)

In [209]:
#для тестовой выборки
print_regression_metrics(y_valid, y_valid_pred)

MSE = 29.02, RMSE = 5.39


In [210]:
#для тренировочной выборки
print_regression_metrics(y_train, y_train_pred)

MSE = 20.54, RMSE = 4.53


In [219]:
#Реализуем функцию вычисления градиента функции  и шаг градиентного спуска:
def calc_mse_gradient(X, y, theta):
    n = X.shape[0]
    y_hat = X.dot(theta)
    grad = 1. / n * X.transpose().dot(y_hat - y)
    return grad

def gradient_step(theta, theta_grad, alpha):
    return theta - alpha * theta_grad

In [220]:
#Сделаем процедуру оптимизации:
def optimize(X, y, grad_func, start_theta, alpha, n_iters):
    theta = start_theta.copy()
    
    for i in range(n_iters):
        theta_grad = grad_func(X, y, theta)
        theta = gradient_step(theta, theta_grad, alpha)
    return theta

In [221]:
#Оптимизируем параметр линейной регрессии  на всех данных:
m = X.shape[1]
theta = optimize(X, y, calc_mse_gradient, np.ones(m), 0.001, 100)
np.linalg.norm(theta) 

inf

#Градиент получился очень большой, у нас вышло переполнение, и  выдала значения nan. Для того чтобы этого избежать, нужно применить стандартную нормализацию:


In [222]:
X, y = data['data'], data['target']
X_norm = (X - X.mean(axis=0)) / X.std(axis=0)
y_norm = (y - y.mean(axis=0)) / y.std(axis=0)
#нормализуем
e = np.ones(X_norm.shape[0])[: , np.newaxis]
X_norm = np.hstack([e, X_norm])

In [223]:
theta = optimize(X_norm, y, calc_mse_gradient, np.ones(X_norm.shape[1]), 0.01, 10000)
np.linalg.norm(theta) 

23.698188183235512

In [224]:
#Сделаем предсказание, посчитаем значение ошибок:
y_pred = X_norm.dot(theta)
print_regression_metrics(y, y_pred)

MSE = 21.89, RMSE = 4.68


In [225]:
#Разобьём выборку на train/valid, оптимизируем , сделаем предсказания и посчитаем ошибки MSE и RMSE:
X_train, X_valid, y_train, y_valid = train_test_split(X_norm, y, test_size=0.2)
theta = optimize(X_train, y_train, calc_mse_gradient, np.ones(X_norm.shape[1]), 0.01, 5000)
y_pred = X_valid.dot(theta)

print_regression_metrics(y_valid, y_pred)

MSE = 40.23, RMSE = 6.34


In [218]:
np.linalg.norm(theta) 

23.750440913969214

# Задание 3.5.1

2.0/2.0 points (graded)
Реализуйте матричную линейную регрессию. Какой получился RMSE?
Ответ округлите до сотых, пример ввода: 0.55

In [241]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_boston
from sklearn.datasets import fetch_california_housing
from sklearn.metrics import mean_squared_error, f1_score, accuracy_score, roc_curve, roc_auc_score
from sklearn.model_selection import train_test_split

from matplotlib import pyplot as plt

#Theta function
def linreg_linear(X, y):
    xTrans = X.T
    matrix = xTrans.dot(X)
    matrix = np.linalg.inv(matrix)
    matrix = matrix.dot(xTrans)
    matrix = matrix.dot(y)
    return matrix

#Подготавливаем данные и вычисляем параметр :
data = load_boston()
dataCalifornia = fetch_california_housing()
X, y = data['data'], data['target']
# create ones array  [1,1,1,1 ...] from X rows and create from it new column 
e = np.ones(X.shape[0])[:, np.newaxis]
# Join a sequence of arrays along a new axis.
X = np.hstack([e, X])

theta = linreg_linear(X, y)

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets inclu

In [242]:
# Сделаем предсказание и посчитаем метрики:
y_pred = X.dot(theta)

def print_regression_metrics(y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    print(f'MSE = {mse:.2f}, RMSE = {rmse:.2f}')
#для полноразмерного обучения -полной выборки
print_regression_metrics(y, y_pred)



MSE = 21.89, RMSE = 4.68


### ответ RMSE = 4.68

# Задание 3.5.2

2.0/2.0 points (graded)
Постройте модель при помощи sklearn. Используйте параметры по умолчанию, обучите на всей выборке и посчитайте RMSE.
Ответ округлите до сотых, пример ввода: 0.55

In [243]:
from sklearn.linear_model import LinearRegression

In [249]:
reg = LinearRegression().fit(X, y)
reg.score(X, y)
y_pred = reg.predict(X)
print_regression_metrics(y, y_pred)

MSE = 21.89, RMSE = 4.68


In [247]:
reg.coef_

array([ 0.00000000e+00, -1.08011358e-01,  4.64204584e-02,  2.05586264e-02,
        2.68673382e+00, -1.77666112e+01,  3.80986521e+00,  6.92224640e-04,
       -1.47556685e+00,  3.06049479e-01, -1.23345939e-02, -9.52747232e-01,
        9.31168327e-03, -5.24758378e-01])

In [248]:
theta

array([ 3.64594884e+01, -1.08011358e-01,  4.64204584e-02,  2.05586264e-02,
        2.68673382e+00, -1.77666112e+01,  3.80986521e+00,  6.92224640e-04,
       -1.47556685e+00,  3.06049479e-01, -1.23345939e-02, -9.52747232e-01,
        9.31168327e-03, -5.24758378e-01])

### ответ 4,68

# Задание 3.5.3

2.0/2.0 points (graded)
Чему равно наибольшее стандартное отклонение у признаков?

Ответ округлите до сотых. Пример ввода: 5.55

In [253]:
np.std(X, axis=0).max()

168.3704950393814

### ответ 168.37

# Задание 3.5.4

2.0/2.0 points (graded)
Обучите регрессию без дополнительного столбца единиц. Какой получился RMSE?
Ответ округлите до сотых. Пример ввода: 5.55.

In [261]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_boston
from sklearn.datasets import fetch_california_housing
from sklearn.metrics import mean_squared_error, f1_score, accuracy_score, roc_curve, roc_auc_score
from sklearn.model_selection import train_test_split

from matplotlib import pyplot as plt

#Theta function
def linreg_linear(X, y):
    xTrans = X.T
    matrix = xTrans.dot(X)
    matrix = np.linalg.inv(matrix)
    matrix = matrix.dot(xTrans)
    matrix = matrix.dot(y)
    return matrix

#Подготавливаем данные и вычисляем параметр :
data = load_boston()
dataCalifornia = fetch_california_housing()
X, y = data['data'], data['target']

theta = linreg_linear(X, y)

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets inclu

In [262]:
# Сделаем предсказание и посчитаем метрики:
y_pred = X.dot(theta)

def print_regression_metrics(y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    print(f'MSE = {mse:.2f}, RMSE = {rmse:.2f}')
#для полноразмерного обучения -полной выборки
print_regression_metrics(y, y_pred)

MSE = 24.17, RMSE = 4.92


### ответ 4.68 (ошибка курса)

# Задание 3.5.5

2.0/2.0 points (graded)
Очистите данные от строк, где значение признака  В меньше 50. Какой получился RMSE?
Ответ округлите до сотых, пример ввода: 5.55

In [278]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_boston
from sklearn.datasets import fetch_california_housing
from sklearn.metrics import mean_squared_error, f1_score, accuracy_score, roc_curve, roc_auc_score
from sklearn.model_selection import train_test_split

from matplotlib import pyplot as plt

#Theta function
def linreg_linear(X, y):
    xTrans = X.T
    matrix = xTrans.dot(X)
    matrix = np.linalg.inv(matrix)
    matrix = matrix.dot(xTrans)
    matrix = matrix.dot(y)
    return matrix

#Подготавливаем данные и вычисляем параметр :
data = load_boston()
dataCalifornia = fetch_california_housing()
X, y = data['data'][data['data'][:, 11]>=50], data['target'][data['data'][:, 11]>=50]
# create ones array  [1,1,1,1 ...] from X rows and create from it new column 
e = np.ones(X.shape[0])[:, np.newaxis]
# Join a sequence of arrays along a new axis.
X = np.hstack([e, X])
theta = linreg_linear(X, y)

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets inclu

In [281]:
# Сделаем предсказание и посчитаем метрики:
y_pred = X.dot(theta)

def print_regression_metrics(y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    print(f'MSE = {mse:.2f}, RMSE = {rmse:.2f}')
#для полноразмерного обучения -полной выборки
print_regression_metrics(y, y_pred)

MSE = 21.79, RMSE = 4.67


### ответ 4.68 (ошибка курса)

# Задание 3.5.6

2.0/2.0 points (graded)
Нормализуйте признаки и обучите линейную регрессию матричным методом. Какой получился RMSE?
Ответ округлите до сотых, пример ввода: 5.55

In [299]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_boston
from sklearn.datasets import fetch_california_housing
from sklearn.metrics import mean_squared_error, f1_score, accuracy_score, roc_curve, roc_auc_score
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from matplotlib import pyplot as plt

#Theta function
def linreg_linear(X, y):
    xTrans = X.T
    matrix = xTrans.dot(X)
    matrix = np.linalg.inv(matrix)
    matrix = matrix.dot(xTrans)
    matrix = matrix.dot(y)
    return matrix

#Подготавливаем данные и вычисляем параметр :
data = load_boston()
dataCalifornia = fetch_california_housing()
X, y = data['data'], data['target']
#X_norm = (X - X.mean(axis=0)) / X.std(axis=0)
X_norm = MinMaxScaler().fit_transform(X, y)
y_norm = (y - y.mean(axis=0)) / y.std(axis=0)

theta = linreg_linear(X_norm, y)

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets inclu

In [300]:
# Сделаем предсказание и посчитаем метрики:
y_pred = X_norm.dot(theta)

def print_regression_metrics(y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    print(f'MSE = {mse:.2f}, RMSE = {rmse:.2f}')
#для полноразмерного обучения -полной выборки
print_regression_metrics(y, y_pred)

MSE = 27.47, RMSE = 5.24


In [305]:
#вариант 2
X2 = pd.DataFrame(X)
def normalization_df(X):
    result = X.copy()
    for column_name in X.columns:
        max_value = X[column_name].max()
        min_value = X[column_name].min()
        std_value = X[column_name].mean()
        if max_value == min_value:
            result[column_name] = 1
        else: 
            result[column_name] = (X[column_name] - min_value) / (max_value - min_value)
    return result
X_norm2 = normalization_df(X2)
theta2 = linreg_linear(X_norm2, y)

# Сделаем предсказание и посчитаем метрики:
y_pred = X_norm2.dot(theta2)

def print_regression_metrics(y_true, y_pred):
    mse = mean_squared_error(y_true, y_pred)
    rmse = np.sqrt(mse)
    print(f'MSE = {mse:.2f}, RMSE = {rmse:.2f}')
#для полноразмерного обучения -полной выборки
print_regression_metrics(y, y_pred)

MSE = 27.47, RMSE = 5.24


### ответ 4.68 (ошибка курса)